libsolidity/formal/SMTEncoder.cpp

```diff
@@ -1227,39 +1227,64 @@ pair<smtutil::Expression, smtutil::Expression> SMTEncoder::arithmeticOperation(
 	else
 		intType = TypeProvider::uint256();
 
-	smtutil::Expression valueNoMod(
-		_op == Token::Add ? _left + _right :
-		_op == Token::Sub ? _left - _right :
-		_op == Token::Div ? division(_left, _right, *intType) :
-		_op == Token::Mul ? _left * _right :
-		/*_op == Token::Mod*/ _left % _right
-	);
+	optional<smtutil::Expression> valueUnbounded;
+	if (_op == Token::Add)
+		valueUnbounded = _left + _right;
+	else if (_op == Token::Sub)
+		valueUnbounded = _left - _right;
+	else if (_op == Token::Mul)
+		valueUnbounded = _left * _right;
+	else if (_op == Token::Div)
+		valueUnbounded = division(_left, _right, *intType);
+	else /* _op == Token::Mod */
+		valueUnbounded = _left % _right;
 
 	if (_op == Token::Div || _op == Token::Mod)
+	{
 		m_context.addAssertion(_right != 0);
 
+		// mod and unsigned division never underflow/overflow
+		if (_op == Token::Mod || !intType->isSigned())
+			return {*valueUnbounded, *valueUnbounded};
+
+		// The only case where division overflows is
+		// - type is signed
+		// - LHS is type.min
+		// - RHS is -1
+		// the result is then -(type.min), which wraps back to type.min
+		smtutil::Expression maxLeft = _left == smt::minValue(*intType);
+		smtutil::Expression minusOneRight = _right == -1;
+		smtutil::Expression wrap = smtutil::Expression::ite(maxLeft && minusOneRight, smt::minValue(*intType), *valueUnbounded);
+		return {wrap, *valueUnbounded};
+	}
+
 	auto symbMin = smt::minValue(*intType);
 	auto symbMax = smt::maxValue(*intType);
 
 	smtutil::Expression intValueRange = (0 - symbMin) + symbMax + 1;
+	string suffix = to_string(_operation.id()) + "_" + to_string(m_context.newSlackId());
+	smt::SymbolicIntVariable k(intType, intType, "k_" + suffix, m_context);
+	smt::SymbolicIntVariable m(intType, intType, "m_" + suffix, m_context);
```

Comment:
I'm not quite sure how variable namespaces are handled in all the smt code - we can be sure this doesn't collide with anything ever and we won't forget that those names exist? But should probably be fine.

Reply:
Yea, the namespace `smt` is all inside `libsolidity/formal`. The namespace from `libsmtutil` is `smtutil`.

Reply:
I meant the "namespace" in names of SMT variables :-). I.e. the fact the we reserve the rather generic prefices ``k_`` and ``m_`` for these here now.

Reply:
Ah right. I think it should be fine. Even if there are vars called `k` or `m`, they won't have the same AST id of the binary operation